pytorch_lightning/callbacks/stochastic_weight_avg.py

```diff
@@ -280,3 +295,34 @@ def avg_fn(
     ) -> torch.FloatTensor:
         """Adapted from https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L95-L97."""
         return averaged_model_parameter + (model_parameter - averaged_model_parameter) / (num_averaged + 1)
+
+    def on_save_checkpoint(
+        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
+    ) -> dict:
+        return {
+            "n_averaged": 0 if self.n_averaged is None else self.n_averaged.item(),
+            "scheduler_step_count": None if self._swa_scheduler is None else self._swa_scheduler._step_count,
+            "average_model_parameters": self._get_average_model_parameters(trainer),
+        }
+
+    def on_load_checkpoint(
+        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", callback_state: Dict[str, Any]
+    ) -> None:
+        self._init_n_averaged = callback_state["n_averaged"]
+        self._scheduler_step_count = callback_state["scheduler_step_count"]
+        self._load_average_model_parameters(callback_state["average_model_parameters"])
+
```

Comment:
since on_save and on_load_checkpoint were not implemented in older versions, it would be good to raise an error or provide a default behavior if the keys are no in the checkpoint (when loading a checkpoint that was saved with an older Lightning version). Right now, it would result in a `KeyError`. What are your thoughts here?

Reply:
I believe this method will never be called if we load a checkpoint saved from an older version, as `TrainerCallbackHookMixin` checks that the callback data is truthy before calling this `on_load_checkpoint`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/06c5903600b8ed288757a71a00d686120a5fd7a7/pytorch_lightning/trainer/callback_hook.py#L82-L86
(and it only saves something to the checkpoint if it gets a truthy value back from `on_save_checkpoint`).

So it doesn't seem like there's much that can be done from the callback's point of view to handle checkpoints from an older version without data, we'll just silently continue without realising we're restarting without any saved state. Maybe we could detect this based on `trainer.current_epoch` when starting a training epoch? As things are, if we restart after the SWA starting epoch without any checkpoint data, we'll now reinitialize SWA when we wouldn't have before, but `n_averaged` and the scheduler step count will start at zero.

Reply:
@awaelchli, I've done some testing to verify this behaviour. If loading an old checkpoint without SWA data everything works fine if the checkpoint was from before the SWA start epoch. If loading after the start epoch then the behaviour will be a bit different than if checkpoint data was loaded but there isn't a crash. I think this is an improvement on what currently happens on master where SWA isn't reinitialized at all, but it might be nice to log a warning about this?

Reply:
I've added that warning now. Happy to change the approach if you disagree though.

Looks like the release of PyTorch 1.10.1 has caused the build to start failing.

Reply:
thanks, looks good :)